website/addons/osfstorage/model.py

```diff
@@ -8,104 +8,62 @@
 import pymongo
 
 from modularodm import fields, Q
-from modularodm import exceptions as modm_errors
-from modularodm.storage.base import KeyExistsException
 from dateutil.parser import parse as parse_date
+from modularodm.exceptions import NoResultsFound
 
 from framework.auth import Auth
 from framework.mongo import StoredObject
 from framework.analytics import get_basic_counters
+
 from website.models import NodeLog
 from website.addons.base import AddonNodeSettingsBase, GuidFile
 
 from website.addons.osfstorage import logs
+from website.addons.osfstorage import utils
 from website.addons.osfstorage import errors
 from website.addons.osfstorage import settings
 
 
 logger = logging.getLogger(__name__)
 
-oid_primary_key = fields.StringField(
-    primary=True,
-    default=lambda: str(bson.ObjectId())
-)
-
-
-def copy_file_tree(tree, node_settings):
-    """Recursively copy file tree.
-
-    :param OsfStorageFileTree tree: Tree to copy
-    :param node_settings: Root node settings record
-    """
-    children = [copy_files(child, node_settings) for child in tree.children]
-    clone = tree.clone()
-    clone.children = children
-    clone.node_settings = node_settings
-    clone.save()
-    return clone
-
-
-def copy_file_record(record, node_settings):
-    """Copy versions of an `OsfStorageFileRecord`. Versions are copied
-    by primary key and will not be duplicated in the database.
-
-    :param OsfStorageFileRecord record: Record to copy
-    :param node_settings: Root node settings record
-    """
-    clone = record.clone()
-    clone.versions = record.versions
-    clone.node_settings = node_settings
-    clone.save()
-    return clone
-
-
-def copy_files(files, node_settings):
-    if isinstance(files, OsfStorageFileTree):
-        return copy_file_tree(files, node_settings)
-    if isinstance(files, OsfStorageFileRecord):
-        return copy_file_record(files, node_settings)
-    raise TypeError('Input must be `OsfStorageFileTree` or `OsfStorageFileRecord`')
-
 
 class OsfStorageNodeSettings(AddonNodeSettingsBase):
-
+    complete = True
+    has_auth = True
+    root_node = fields.ForeignField('OsfStorageFileNode')
     file_tree = fields.ForeignField('OsfStorageFileTree')
 
-    @property
-    def has_auth(self):
-        return True
+    def on_add(self):
+        if self.root_node:
+            return
 
-    @property
-    def complete(self):
-        return True
+        self.save()
```

Comment:
Erm... unless the the record needs to be save in order for the root node's backref to be saved properly (?).

If the save *is* necessary, please add a line comment explaining why it's necessary.